### ISSUE TITLE
zsh-autocomplete breaks p10k transient prompt

### ISSUE THREAD

**Issue by yrammos** [marlonrichert/zsh-autocomplete#173]
@marlonrichert excuse me for a rather elliptical issue report. The "[transient prompt](https://github.com/romkatv/powerlevel10k#transient-prompt)" feature of Powerlevel10k works only once, and then stops, if `zsh-autocomplete` is included in the configuration. This wouldn't be the end of the world, as far as I'm concerned, but I'm worried it might be a symptom of an underlying low-level or design issue, with unexpected consequences.

I'd appreciate your advice on how to proceed, if at all. This might not be a "bug" but an incompatibility problem between the two plugins involved. Semantics aside, here are two extraordinary plugins by different authors, and it would be nice to be able to run them in parallel.

(I faced a similar conflict between your `zsh-snap` and `p10k` two days ago, which would unpredictably cause zsh to seg-fault. Running low on time, I didn't report that one, instead refactoring my `.zshrc` to source plugins directly. The coexistence of the two lazy-loading mechanisms, yours and his, may not be harmonious.)

Thanks for your time.

**Comment by marlonrichert** [comment]
> (I faced a similar conflict between your `zsh-snap` and `p10k` two days ago, which would unpredictably cause zsh to seg-fault. Running low on time, I didn't report that one, instead refactoring my `.zshrc` to source plugins directly. The coexistence of the two lazy-loading mechanisms, yours and his, may not be harmonious.)

I would appreciate if you could open an issue on that, too. I'm not sure what lazy-loading mechanism you are referring to.

**Comment by marlonrichert** [comment]
> excuse me for a rather elliptical issue report. The "[transient prompt](https://github.com/romkatv/powerlevel10k#transient-prompt)" feature of Powerlevel10k works only once, and then stops, if `zsh-autocomplete` is included in the configuration. This wouldn't be the end of the world, as far as I'm concerned, but I'm worried it might be a symptom of an underlying low-level or design issue, with unexpected consequences.
> 
> I'd appreciate your advice on how to proceed, if at all. This might not be a "bug" but an incompatibility problem between the two plugins involved. Semantics aside, here are two extraordinary plugins by different authors, and it would be nice to be able to run them in parallel.

I'm able to run both of them in parallel without issue. Can you please try the following?
```zsh
> cd "$(mktemp -d)"  # Create a temp dir and enter it.
> ZDOTDIR=$PWD HOME=$PWD zsh -f  # Start a subshell without config files.
% source path/to/zsh-autocomplete.plugin.zsh; source path/to/powerlevel10k.zsh-theme; source path/to/.p10k.zsh
```
If that works for you, too, then the problem is elsewhere in your Zsh config.

**Comment by euglevi** [comment]
I have this problem too. It is not a big deal but it would be nice if it was solved.

I tried with the subshell in the temp dir. The "transient prompt" feature of powerlevel10k is still not working. It is not a problem with p10k in general, just with this feature.

By the way, thank you very much for your wonderful plugins!

**Comment by yrammos** [comment]
@marlonrichert thank you for the speedy reply. In fact the problem persists with the minimal shell of your instructions.

$ZSH_PATCHLEVEL: `zsh-5.8-0-g77d203f`

Plugin commit: `db3370d4cf3223029a7a5e3a06ba6f529fe5fa77`

**Comment by yrammos** [comment]
> I would appreciate if you could open an issue on that, too, in the Znap repo. I'm not sure what lazy-loading mechanisms you are referring to.

(On this tangent: I will file the issue as soon as time allows me to revert to `znap` and reproduce it. Thanks as ever.)

**Comment by marlonrichert** [comment]
@yrammos @euglevi In what way does the transient prompt stop working? Does it simply become non-transient? Or does it break somehow?

**Comment by yrammos** [comment]
It simply becomes non-transient.

**Comment by euglevi** [comment]
@yrammos the same for me. No message error either in autocomplete or in p10k and both seem to work fine on everything else.

**Comment by marlonrichert** [comment]
OK, I can see it now when I try. I have no idea what causes this or how to fix it. You might have better luck filing an issue with Powerlevel10k.

**Comment by romkatv** [comment]
This issue is caused by zsh-autocomplete removing zle-line-finished hooks from all other plugins. Powerlevel10k installs a zle-line-finished hook and then zsh-autocomplete removes it. To reproduce:

```text
% zsh -f
% git clone --depth=1 https://github.com/marlonrichert/zsh-autocomplete.git
% zle-line-finish() print '\nzle-line-finish'; zle -N zle-line-finish
% source zsh-autocomplete/zsh-autocomplete.plugin.zsh
zle-line-finish

% true
%
```

Notice how "zle-line-finish" stops being printed once zsh-autocomplete is loaded. Let's see how the hook looks like:

```text
% echo $widgets[zle-line-finish]
user:azhw:zle-line-finish
% functions azhw:zle-line-finish
azhw:zle-line-finish () {
        local -a hook_widgets
        local hook
        zstyle -a $WIDGET widgets hook_widgets
        for hook in "${(@)${(@on)hook_widgets[@]}#<->:}"
        do
                if [[ "$hook" = user:* ]]
                then
                        zle "$hook" -N -- "$@"
                else
                        zle "$hook" -Nw -- "$@"
                fi || return
        done
        return 0
}
% zstyle -a zle-line-finish widgets hooks && typeset -p hooks
typeset -a hooks=( 1:.autocomplete.key.line-finish 2:.autocomplete.async.clear )
% functions .autocomplete.key.line-finish
.autocomplete.key.line-finish () {
        echoti rmkx
        return 0
}
% functions .autocomplete.async.clear
.autocomplete.async.clear () {
        zle -Rc
        .autocomplete.async.stop
        .autocomplete.async.reset-context
        return 0
}
% functions .autocomplete.async.stop
.autocomplete.async.stop () {
        .autocomplete.async.kill _autocomplete__async_complete_fd _autocomplete__async_complete_pid
        .autocomplete.async.kill _autocomplete__async_timeout_fd _autocomplete__async_timeout_pid
        return 0
}
% functions .autocomplete.async.reset-context
.autocomplete.async.reset-context () {
        typeset -gH curcontext=''
        zstyle -s :autocomplete: default-context curcontext
        return 0
}
```

As you can see, nothing in there invokes the original zle-line-finish hook.

**Comment by marlonrichert** [comment]
Fixed.

**Comment by yrammos** [comment]
Thank you, @marlonrichert.

**Comment by eddyg** [comment]
Thank you, Roman! 👌

**Comment by yrammos** [comment]
@marlonrichert I'm afraid I rushed to confirm. In fact, the problem doesn't go away on my end, using your latest update and the same minimal shell you provided above.

**Comment by marlonrichert** [comment]
@yrammos Hm, you're right. The test case that Roman provided, though, now succeeds:
```zsh
% zsh -f
% git clone --depth=1 https://github.com/marlonrichert/zsh-autocomplete.git
% zle-line-finish() print '\nzle-line-finish'; zle -N zle-line-finish
% source zsh-autocomplete/zsh-autocomplete.plugin.zsh
zle-line-finish

% true
zle-line-finish

%
```

**Comment by euglevi** [comment]
After the yesterday's update, my transient prompt started working. The problem seems to be solved somehow. Is it the same for you @yrammos?

**Comment by marlonrichert** [comment]
@euglevi Can you do [`git bisect`](https://git-scm.com/docs/git-bisect) to find the exact commit that fixed the issue?

**Comment by yrammos** [comment]
Not confirming — unfortunately the issue persists.

**Comment by euglevi** [comment]
@marlonrichert I have tried `git bisect`, but with no success. It goes up to the commit that I know for sure was bad but I experience no problem now. 

Funny enough, when I try with the minimal configuration it still experiences problems. So, there must be something in my .zshrc  that solves the problem, I guess. I have tried to find what it is, but with no luck up to now.

**Comment by marlonrichert** [comment]
@yrammos @eddyg @euglevi Can you paste this into your command line and see if it fixes the problem?
```zsh
.autocomplete.async.clear() {
  zle -M ''
  .autocomplete.async.stop
  .autocomplete.async.reset-context
  return 0
}
```

**Comment by yrammos** [comment]
@marlonrichert indeed this seems to work!

**Comment by marlonrichert** [comment]
@yrammos @eddyg @euglevi Fix committed and pushed in.

**Comment by euglevi** [comment]
Sorry for the slow answer...busy day. After the update, for me the prompt is broken. I tried with the minimal configuration, and as soon as I source p10k the left side of the prompt just disappears. This is before I source my specific configuration of p10k. Before the last commit zsh-autocomplete was doing very well.

**Comment by marlonrichert** [comment]
@euglevi Sounds like a different problem. Can you open a new issue for it, please?

**Comment by marlonrichert** [comment]
I had to revert the fix because of #175.

@romkatv Any help? Fixing your test case did not fix the actual problem.

**Comment by romkatv** [comment]
I’m afraid I don’t have the time to debug this.

**Comment by marlonrichert** [comment]
@romkatv Never mind, I found the culprit and filed an issue P10k: https://github.com/romkatv/powerlevel10k/issues/1238 @yrammos @eddyg @euglevi

**Comment by euglevi** [comment]
Thank you @marlonrichert. Again, not sure why with my .zshrc (with Zinit) transient prompt is actually working for me at the moment. This eventually means that there is no deep incompatibility.

**Comment by marlonrichert** [comment]
@yrammos @eddyg @euglevi It should be fixed now. Thank you for bearing with me.

**Comment by yrammos** [comment]
@marlonrichert, thank you very much.
